### PR TITLE
Support per-target linker scripts for tests

### PR DESCRIPTION
**Pull request by keith-packard**
qemu for arm doesn't have a single machine which supports all of the desired cpu targets, instead we need to use a range of them with different memory configurations.
    
To make this work with a multilib build, add support for cross compilation properties for each qemu machine that are mapped to the relevant gcc targets.

### CODE REVIEW

File: scripts/run-arm [hunk 1]
```diff
@@ -45,154 +45,187 @@ machine=unknown
...
+                ;;
+        esac
+        ;;
+    v8.1-M.mainline/Microcontroller)
+        case "$fp_arch" in
+            FPv5/FP-D16)
+                case "$fp_use" in
+                    SP)
+                        cpu=cortex-m55
+                        ;;
```
Comment (domin144) [inline scripts/run-arm:121]: I can see cortex-m55 is used for double precision case in "v8-M.*". Cannot it be used here for both single and double precision?

File: newlib/libc/tinystdio/atod_engine.c [hunk 2]
```diff
@@ -37,13 +37,16 @@ __atod_engine(uint64_t u64, int exp)
...
+    if (exp < 0) {
+        do {
+            flt *= 1e-1;
+            exp++;
+        } while (exp < 0);
+    } else if (exp > 0) {
+        do {
+            flt *= 1e1;
+            exp--;
+        } while (exp > 0);
```
Comment (domin144) [inline newlib/libc/tinystdio/atod_engine.c:49]: Looks similar to the bug hit in dirname. I decided to skip the test in `run-arm` for that case.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110791

File: scripts/run-arm [hunk 3]
```diff
@@ -45,154 +45,187 @@ machine=unknown
...
-    *thumb_v7_fp_hard*)
-	cpu=cortex-a8
-	;;
-    *thumb_v7_fp_softfp*)
-	cpu=cortex-a8
-	;;
+cpu_arch=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_CPU_arch:/ { print $2 }'`
+cpu_profile=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_CPU_arch_profile:/ { print $2 }'`
+fp_arch=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_FP_arch/ { print $2 }'`
+fp_use=`arm-none-eabi-readelf -A "$elf" | awk '/Tag_ABI_HardFP_use/ { print $2 }'`
```
Comment (domin144) [inline scripts/run-arm:54]: Neat. I did't know such details can be read from the 'ELF' file.

**Review by domin144 — COMMENTED** [review]
Great work done here. All platforms tested and with no virtm patch to QEMU!

File: picocrt/machine/arm/crt0.c [hunk 4]
```diff
@@ -104,20 +104,18 @@ _start(void)
 	/* Initialize stack pointer */
 	__asm__("mov sp, %0" : : "r" (__stack));
 
-#if __ARM_ARCH == 7
-#ifdef __thumb__
+#if defined(__thumb2__)
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply picocrt/machine/arm/crt0.c:107]: yup, thanks.

File: newlib/testsuite/newlib.string/tstring.c [hunk 5]
```diff
@@ -56,14 +56,17 @@ void myset (char *target, char ch, int size)
...
-static char buffer2[MAX_1];
-static char buffer3[MAX_1];
-static char buffer4[MAX_1];
-static char buffer5[MAX_2];
-static char buffer6[MAX_2];
-static char buffer7[MAX_2];
-static char expected[MAX_1];
+char *getbuf(size_t size, const char *template)
+{
+    char *buf = calloc(size, 1);
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/testsuite/newlib.string/tstring.c:61]: nah, it's allocated at the top of `main`, so it happens exactly once during program execution.

File: scripts/cross-arm-none-eabi.txt [hunk 6]
```diff
@@ -19,10 +19,44 @@ endian = 'little'
 
 [properties]
 skip_sanity_check = true
-separate_boot_flash = true
-default_boot_flash_addr = '0x00000000'
-default_boot_flash_size = '0x00000400'
-default_flash_addr = '0x00000400'
-default_flash_size = '0x003ffc00'
+default_flash_addr = '0x00000000'
+default_flash_size = '0x00400000'
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply scripts/cross-arm-none-eabi.txt:23]: Yup, and I've got lots of those in use. The separate configuration also uses more flash as it constrains what can live in the lowest portion. I think we'll leave these alone.

File: scripts/run-arm (hunk 1 above)
Comment (keith-packard) [reply scripts/run-arm:121]: One of the goals here is to make sure the code doesn't use features not found in the target architecture, so limiting the core to something that more precisely fits the compiler settings is why I have stuff like this. In this case, I believe the m33 is a v8 core with SP float while the m55 is a v8.1 core with SP/DP float. Or I could just be wrong :-)

File: scripts/run-arm (hunk 3 above)
Comment (keith-packard) [reply scripts/run-arm:54]: yeah, someone else provided hacks to the `run-riscv` script that did this. I've applied it to run-arc (where it was dead simple) and now to run-arm (where it's a bit complicated).

File: meson.build [hunk 7]
```diff
@@ -992,6 +900,165 @@ else
...
+				output: picolibcpp_ld_file,
+				configuration: picolibcpp_ld_data,
+				install: picolibc_ld_install,
+				install_dir: lib_dir))
+    endif
+		   
+    set_variable(picolibcpp_ld_variable, get_variable(picolibcpp_ld_config_variable))
+
+  endif
+endforeach
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply meson.build:1060]: It's probably way too complicated, given that it's only for tests. But, it got the job done.

File: meson.build [hunk 8]
```diff
@@ -992,6 +900,165 @@ else
...
+    # (ALIGN_WITH_INPUT, which is not supported by ld.lld) and ld.lld
+    # (ALIGN(__tls_align), which is rejected as non-constant by ld.bfd), so we use
+    # the {BFD,LLD}_{START,END} templates to comment out the incompatible flag.
+    if cc.get_linker_id() == 'ld.lld'
+      picolibc_linker_type_data.set('BFD_START', '/* For ld.bfd: ')
+      picolibc_linker_type_data.set('BFD_END', '*/')
+      picolibc_linker_type_data.set('LLD_START', '')
+      picolibc_linker_type_data.set('LLD_END', '')
+      if host_cpu_family == 'riscv'
+	# ld.lld before version 15 did not support linker relaxations, disable
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply meson.build:922]: good plan. I've updated .editorconfig to match.

File: newlib/libc/tinystdio/atod_engine.c (hunk 2 above)
Comment (keith-packard) [reply newlib/libc/tinystdio/atod_engine.c:49]: Oh, now I remember why I didn't -- this bug doesn't just cause the test to fail, it causes it to infinite loop, which we can't fix with hacks in run-arm. I've made the work-around simpler *and* made it specific to GCC 12.2.1 with -Os.

Comment (keith-packard) [reply newlib/libc/tinystdio/atod_engine.c:49]: oh, and applied the same fix to `__atof_engine`

File: picocrt/machine/arm/crt0.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (keith-packard) [reply picocrt/machine/arm/crt0.c]: Nope, we don't need to know if the target *supports* thumb2, we need to know if the compiler is generating thumb2 instructions. The only way I know how to do that is with `#ifdef __thumb2__`. Open to other suggestions, ofc.

File: scripts/run-arm (hunk 1 above)
Comment (domin144) [reply scripts/run-arm:121]: I agree it's better to use cortex-m33 for SP in 8m case. I just wonder why, for the 8.1m case, m55 is used only for SP and DP is left unhandled. I thought m55 should be able to handle DP case, too.
Anyway, gcc 12.2 does not provide a multilib variant for v8.1m+fp.dp. As there would be no test to confirm the configuration works, maybe it's better to leave it as is for now.

File: picocrt/machine/arm/crt0.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (domin144) [reply picocrt/machine/arm/crt0.c]: Yes, it has to stay `__thumb2__` then.

File: newlib/libc/tinystdio/atod_engine.c (hunk 2 above)
Comment (domin144) [reply newlib/libc/tinystdio/atod_engine.c]: I like the solution with changing the optimization mode. It is much nicer, than ignoring the bug by skipping the test or hacking nice code.

File: scripts/run-arm (hunk 1 above)
Comment (keith-packard) [reply scripts/run-arm:121]: Oh, now I see your point. It looks like cortex-m55 could handle all of the v8.1m cases, even those not currently part of the gcc multilib configuration? Changing that would simplify the script, which seems like a good idea.

**Pull request merged**